File: picasso-commons/src/com/hendraanggrian/picasso/commons/transformation/ColorGrayscaleTransformer.java
package com.hendraanggrian.picasso.commons.transformation;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Paint;
import android.os.Bundle;
import android.support.annotation.NonNull;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
class ColorGrayscaleTransformer extends Transformer {

    @NonNull
    @Override
    public Bitmap transform(@NonNull Bitmap source, boolean shouldRecycle) {
        Bitmap target = createDefaultBitmap(source);

        ColorMatrix saturation = new ColorMatrix();
        saturation.setSaturation(0f);
        Paint paint = new Paint();
        paint.setColorFilter(new ColorMatrixColorFilter(saturation));
        new Canvas(target).drawBitmap(source, 0, 0, paint);

        if (shouldRecycle) {
            source.recycle();
        }
        return target;
    }

    @NonNull
    @Override
    protected Bundle keyBundle() {
        Bundle bundle = new Bundle();
        bundle.putString(KEY_NAME, "ColorGrayscaleTransformer");
        return bundle;
    }
}

File: picasso-commons/src/com/hendraanggrian/picasso/commons/transformation/ColorGrayscaleTransformer.kt
package com.hendraanggrian.picasso.commons.transformation

import android.graphics.*
import com.squareup.picasso.Transformation

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
internal class ColorGrayscaleTransformer : Transformation {

    override fun transform(source: Bitmap): Bitmap {
        val target = Bitmap.createBitmap(source.width, source.height, Bitmap.Config.ARGB_8888)
        Canvas(target).drawBitmap(source, 0f, 0f, Paint().apply {
            colorFilter = ColorMatrixColorFilter(ColorMatrix().apply { setSaturation(0f) })
        })
        source.recycle()
        return target
    }

    override fun key() = "ColorGrayscaleTransformer"
}